tests/runtime/servicers/test_training_management_servicer.py

```diff
@@ -75,6 +75,69 @@ def test_training_runs(training_management_servicer, training_pool):
     assert response.state == TrainingStatus.COMPLETED.value
 
 
+def test_training_cannot_cancel_on_completed_training(training_management_servicer):
+    # Create a future and set it in the training manager
+    event = threading.Event()
+    model_future = MODEL_MANAGER.train(
+        SampleModule,
+        DataStream.from_iterable([]),
+        wait_event=event,
+    )
+
+    # send a request, check it's not errored
+    request = TrainingInfoRequest(training_id=model_future.id).to_proto()
+    response = training_management_servicer.GetTrainingStatus(request, context=None)
+    assert response.state != TrainingStatus.ERRORED.value
+
+    event.set()
+    model_future.wait()
+
+    # cancel the training request, but this won't change its status as it was finished
+    training_management_servicer.CancelTraining(request, context=None)
+    response = training_management_servicer.GetTrainingStatus(request, context=None)
+    assert response.state == TrainingStatus.COMPLETED.value
+
+
+def test_training_cancel_on_correct_id(training_management_servicer):
+    # Create a training future for first model with a wait event
+    event_1 = threading.Event()
+    model_future_1 = MODEL_MANAGER.train(
+        SampleModule,
+        DataStream.from_iterable([]),
+        wait_event=event_1,
+    )
+
+    def unblock_training_thread():
+        event_1.set()
+        model_future_1.wait()
+
+    request_1 = TrainingInfoRequest(training_id=model_future_1.id).to_proto()
+    response_1 = training_management_servicer.GetTrainingStatus(request_1, context=None)
+    assert response_1.state == TrainingStatus.RUNNING.value
+
+    # Model 2 has no wait event, should proceed to complete training
+    model_future_2 = MODEL_MANAGER.train(
+        SampleModule,
+        DataStream.from_iterable([1, 2, 3]),
+    )
+
+    # Cancel first training
+    request_1 = TrainingInfoRequest(training_id=model_future_1.id).to_proto()
+    training_management_servicer.CancelTraining(request_1, context=None)
+
+    response_1 = training_management_servicer.GetTrainingStatus(request_1, context=None)
+    assert response_1.state == TrainingStatus.CANCELED.value
+
+    # release the blocked training that is waiting to clean up
```

Comment:
I'm still not sure this is doing what we expect. You shouldn't need to run this in a separate thread and you shouldn't need to do `event_1.set()` since the cancelation _should_ terminate the blocking `wait` in the training function.